File: frontend/src/shared/api/assignments.ts
import type {
  AssignmentDashboard,
  AssignmentExportResponse,
  AssignmentItem,
  AssignmentListResponse,
  AssignmentReport,
  AssignmentUploadArchivePayload,
  RepositoryAnalysisPayload,
} from "../types/assignments";

const apiBaseUrl = import.meta.env.VITE_API_BASE_URL ?? "http://localhost:8000/api";

async function requestJson<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${apiBaseUrl}${path}`, {
    headers: {
      "Content-Type": "application/json",
      ...init?.headers,
    },
    ...init,
  });

  if (!response.ok) {
    throw new Error(`Request failed: ${response.status}`);
  }

  return response.json() as Promise<T>;
}

function authHeaders(token?: string): HeadersInit {
  return token ? { Authorization: `Bearer ${token}` } : {};
}

export function analyzeDemoAssignment(token?: string): Promise<AssignmentReport> {
  return requestJson<AssignmentReport>("/assignments/analyze", {
    method: "POST",
    headers: authHeaders(token),
    body: JSON.stringify({
      assignment_title: "Flask Web 项目实践",
      course_id: "course_web_2026",
      class_id: "class_cs_2024_01",
      student_id: "student_001",
      repository_url: "https://example.edu/demo/flask-project",
      description: "示例作业包含 Flask 路由、SQLite 数据访问、README 和基础测试。",
      files: [
        {
          path: "app.py",
          content:
            "from flask import Flask, render_template\napp = Flask(__name__)\n@app.route('/todos')\ndef todos(): return render_template('todos.html')",
        },
        {
          path: "services/todo_service.py",
          content: "import sqlite3\ndef create_todo(title): return sqlite3.connect('demo.db')",
        },
        { path: "tests/test_app.py", content: "def test_todos_page(): assert True" },
        { path: "requirements.txt", content: "flask\npytest" },
        { path: "README.md", content: "Flask Web 项目实践" },
      ],
    }),
  });
}

export function fetchAssignmentDashboard(token?: string): Promise<AssignmentDashboard> {
  return requestJson<AssignmentDashboard>("/assignments/assignment_flask_mvp/dashboard", {
    headers: authHeaders(token),
  });
}

export function fetchAssignments(token?: string): Promise<AssignmentListResponse> {
  return requestJson<AssignmentListResponse>("/assignments", {
    headers: authHeaders(token),
  });
}

export function fetchAssignmentDashboardById(
  assignmentId: string,
  token?: string,
): Promise<AssignmentDashboard> {
  return requestJson<AssignmentDashboard>(`/assignments/${assignmentId}/dashboard`, {
    headers: authHeaders(token),
  });
}

export function fetchAssignmentReport(
  assignmentId: string,
  studentId: string,
  token?: string,
): Promise<AssignmentReport> {
  return requestJson<AssignmentReport>(`/assignments/${assignmentId}/reports/${studentId}`, {
    headers: authHeaders(token),
  });
}

export function exportAssignmentDashboard(
  assignmentId: string,
  token?: string,
): Promise<AssignmentExportResponse> {
  return requestJson<AssignmentExportResponse>(`/assignments/${assignmentId}/export`, {
    headers: authHeaders(token),
  });
}

export function createAssignment(token?: string): Promise<AssignmentItem> {
  return requestJson<AssignmentItem>("/assignments", {
    method: "POST",
    headers: authHeaders(token),
    body: JSON.stringify({
      assignment_id: "assignment_agent_rag_2026",
      title: "智能体 RAG 应用实践",
      course_id: "course_web_2026",
      class_id: "class_cs_2024_01",
      description: "围绕 RAG 检索、引用展示、对话上下文和工程测试完成一次综合实践。",
      rubric_id: "rubric_agent_rag",
    }),
  });
}

export function analyzeRepositoryAssignment(
  payload: RepositoryAnalysisPayload,
  token?: string,
): Promise<AssignmentReport> {
  return requestJson<AssignmentReport>("/assignments/analyze", {
    method: "POST",
    headers: authHeaders(token),
    body: JSON.stringify({
      assignment_id: payload.assignmentId,
      assignment_title: payload.assignmentTitle,
      course_id: payload.courseId,
      class_id: payload.classId,
      student_id: payload.studentId,
      repository_url: payload.repositoryUrl,
      description: payload.description,
    }),
  });
}

export async function uploadAssignmentArchive(
  payload: AssignmentUploadArchivePayload,
  token?: string,
): Promise<AssignmentReport> {
  const form = new FormData();
  if (payload.assignmentId) form.append("assignment_id", payload.assignmentId);
  form.append("assignment_title", payload.assignmentTitle);
  form.append("archive", payload.archive);
  if (payload.courseId) form.append("course_id", payload.courseId);
  if (payload.classId) form.append("class_id", payload.classId);
  if (payload.studentId) form.append("student_id", payload.studentId);
  if (payload.rubricId) form.append("rubric_id", payload.rubricId);
  if (payload.repositoryUrl) form.append("repository_url", payload.repositoryUrl);
  if (payload.description) form.append("description", payload.description);

  const response = await fetch(`${apiBaseUrl}/assignments/upload-archive`, {
    method: "POST",
    headers: authHeaders(token),
    body: form,
  });

  if (!response.ok) {
    throw new Error(`Request failed: ${response.status}`);
  }

  return response.json() as Promise<AssignmentReport>;
}

File: backend/app/api/routes/assignments.py
from fastapi import APIRouter, Depends, File, Form, Header, HTTPException, UploadFile, status
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.schemas.assignments import (
    AssignmentCreateRequest,
    AssignmentDashboardResponse,
    AssignmentAnalysisRequest,
    AssignmentAnalysisResponse,
    AssignmentItem,
    AssignmentListResponse,
)
from app.services.assignment_service import AssignmentService
from app.services.auth_service import AuthService
from app.services.submission_archive_service import SubmissionArchiveService

router = APIRouter()


@router.get("", response_model=AssignmentListResponse)
def list_assignments(
    authorization: str | None = Header(default=None),
    db: Session = Depends(get_db),
) -> AssignmentListResponse:
    account = AuthService(db).current_account(authorization)
    return AssignmentService(db).list_assignments(account=account)


@router.post("", response_model=AssignmentItem)
def create_assignment(
    payload: AssignmentCreateRequest,
    authorization: str | None = Header(default=None),
    db: Session = Depends(get_db),
) -> AssignmentItem:
    account = AuthService(db).current_account(authorization)
    return AssignmentService(db).create_assignment(payload, account=account)


@router.post("/analyze", response_model=AssignmentAnalysisResponse)
def analyze_assignment(
    payload: AssignmentAnalysisRequest,
    authorization: str | None = Header(default=None),
    db: Session = Depends(get_db),
) -> AssignmentAnalysisResponse:
    account = AuthService(db).current_account(authorization)
    return AssignmentService(db).analyze(payload, account=account)


@router.post("/upload-archive", response_model=AssignmentAnalysisResponse)
async def upload_assignment_archive(
    assignment_id: str | None = Form(default=None),
    assignment_title: str = Form(...),
    course_id: str | None = Form(default=None),
    class_id: str | None = Form(default=None),
    student_id: str | None = Form(default=None),
    rubric_id: str | None = Form(default=None),
    repository_url: str | None = Form(default=None),
    description: str | None = Form(default=None),
    archive: UploadFile = File(...),
    authorization: str | None = Header(default=None),
    db: Session = Depends(get_db),
) -> AssignmentAnalysisResponse:
    if archive.content_type not in {
        "application/zip",
        "application/x-zip-compressed",
        "multipart/x-zip",
        "application/octet-stream",
    }:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="请上传 zip 格式的作业压缩包。",
        )
    archive_bytes = await archive.read()
    try:
        files = SubmissionArchiveService().parse_zip(archive_bytes)
    except ValueError as error:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(error)) from error

    account = AuthService(db).current_account(authorization)
    payload = AssignmentAnalysisRequest(
        assignment_id=assignment_id,
        assignment_title=assignment_title,
        course_id=course_id,
        class_id=class_id,
        student_id=student_id,
        rubric_id=rubric_id,
        repository_url=repository_url,
        description=description,
        files=files,
    )
    return AssignmentService(db).analyze(payload, account=account)


@router.get("/{assignment_id}/reports/{student_id}", response_model=AssignmentAnalysisResponse)
def get_assignment_report(
    assignment_id: str,
    student_id: str,
    authorization: str | None = Header(default=None),
    db: Session = Depends(get_db),
) -> AssignmentAnalysisResponse:
    account = AuthService(db).current_account(authorization)
    return AssignmentService(db).get_report(assignment_id, student_id, account=account)


@router.get("/{assignment_id}/dashboard", response_model=AssignmentDashboardResponse)
def get_assignment_dashboard(
    assignment_id: str,
    authorization: str | None = Header(default=None),
    db: Session = Depends(get_db),
) -> AssignmentDashboardResponse:
    account = AuthService(db).current_account(authorization)
    return AssignmentService(db).get_dashboard(assignment_id, account=account)
